R/types-chr.R

```diff
@@ -0,0 +1,295 @@
+  x <- as_character(x, encoding)
+  stopifnot(is_string(x))
+  x
+}
+#' @rdname as_character
+#' @export
+as_utf8_string <- function(x) {
+  x <- as_utf8_character(x)
+  stopifnot(is_string(x))
+  x
+}
+#' @rdname as_character
+#' @export
+as_native_string <- function(x) {
+  x <- as_native_character(x)
+  stopifnot(is_string(x))
+  x
+}
+
+#' Set encoding of a string or character vector.
+#'
+#' R has specific support for UTF-8 and latin1 encoded strings. This
+#' mostly matters for internal conversions. Thanks to this support,
+#' you can reencode strings to UTF-8 or latin1 for internal
+#' processing, and return these strings without having to convert them
+#' back to the native encoding. However, it is important to make sure
+#' the encoding mark has not been lost in the process. Otherwise the
+#' output will be treated as if encoded according to the current
+#' locale (see \code{\link{set_utf8_locale}()} for documentation about
+#' locale codesets), which is not appropriate if it does not coincide
+#' with the actual encoding. In those situations, you can use these
+#' functions to ensure an encoding mark in your strings.
+#'
+#' @param x A string or character vector.
+#' @param encoding Either an encoding specially handled by R
+#'   (\code{"UTF-8"} or \code{"latin1"}), \code{"bytes"} to inhibit
+#'   all encoding conversions, or \code{"unknown"} if the string
+#'   should be treated as encoded in the current locale codeset.
+#' @seealso \code{\link{set_utf8_locale}()} about the effects of the
+#'   locale, and \code{\link{as_utf8_string}()} about encoding
+#'   conversion.
+#' @export
+#' @examples
+#' # Encoding marks are always ignored on ASCII strings:
+#' str_encoding(str_set_encoding("cafe", "UTF-8"))
+#'
+#' # You can specify the encoding of strings containing non-ASCII
+#' # characters:
+#' cafe <- string(c(0x63, 0x61, 0x66, 0xC3, 0xE9))
+#' str_encoding(cafe)
+#' str_encoding(str_set_encoding(cafe, "UTF-8"))
+#'
+#'
+#' # It is important to consistently mark the encoding of strings
+#' # because R and other packages perform internal string conversions
+#' # all the time. Here is an example with the names attribute:
+#' latin1 <- string(c(0x63, 0x61, 0x66, 0xE9), "latin1")
+#' latin1 <- set_names(latin1)
+#'
+#' # The names attribute is encoded in latin1 as we would expect:
+#' str_encoding(names(latin1))
+#'
+#' # However the names are converted to UTF-8 by the c() function:
+#' str_encoding(names(c(latin1)))
+#' str_bytes(names(c(latin1)))
+#'
+#' # Bad things happen when the encoding marker is lost and R performs
+#' # a conversion. R will assume that the string is encoded according
+#' # to the current locale:
+#' \dontrun{
+#' bad <- set_names(str_set_encoding(latin1, "unknown"))
+#' set_utf8_locale()
+#'
+#' str_encoding(names(c(bad)))
+#' str_bytes(names(c(bad)))
+#' }
+chr_set_encoding <- function(x, encoding = c("unknown", "UTF-8", "latin1", "bytes")) {
+  Encoding(x) <- match.arg(encoding)
```

Comment:
hmm I think that might cause more pain than good, but I'm not sure. `as_utf8_string()` should for sure be idempotent, but in this case maybe it's reasonable to assume that we should only set the encoding when we know it has been lost or was never set (in corner cases like manually created strings)?